### PR TITLE
Fix find_library issue to MacOS Big Sur

### LINKED ISSUE #1582: Couldn't find `libc`'s `posix_memalign` to allocate memory in MacOS Big Sur
In `MacOS Big Sur` dylibs are removed from the disk when they are added to the shared dyld cache. Thus, ctypes.utils.find_libary cannot find the libraries, as we were used to.

Seems that this issue was solved here for cpython 3.9.1 and any patch is necessary.  However it is only available in the release candidate of the Apple Universal Binary version.

Applications that must be compatible with older python versions are writing patches, like this:

```python
elif os.name == "posix" and sys.platform == "darwin":
    from ctypes.macholib.dyld import dyld_find as _dyld_find
    def find_library(name):
        possible = ['@executable_path/../lib/lib%s.dylib' % name,
                    'lib%s.dylib' % name,
                    '%s.dylib' % name,
                    '%s.framework/%s' % (name, name)]
        for name in possible:
            try:
                return _dyld_find(name)
            except ValueError:
                continue
        return None
```

Following the @FabioLuporini suggestion we could implement the same patch like this:

```python
try:
    ... = find_library(...)
except ...:
    # Special cases
    if os.name == "posix" and sys.platform == "darwin":
         ...
    else:
         <suitable error message>
```

### PR DESCRIPTION
**Pull request by speglich**
Closes #1582

### CODE REVIEW

**Comment by mloubout** [comment]
Any idea how this could be tested? don't think gh-action has bigsur

File: devito/data/allocators.py [hunk 1]
```diff
@@ -128,6 +129,10 @@ class PosixAllocator(MemoryAllocator):
     @classmethod
     def initialize(cls):
         handle = find_library('c')
+
+        if handle is None and os.name == "posix" and sys.platform == "darwin":
```
Comment (FabioLuporini) [inline devito/data/allocators.py:133]: Can you add a comment along the lines of:

```
# Special case: on MacOS Big Sur ....
```

Comment (speglich) [reply devito/data/allocators.py:133]: Of course I can, sorry about that!

File: devito/data/allocators.py [hunk 2]
```diff
@@ -128,8 +129,19 @@ class PosixAllocator(MemoryAllocator):
...
+        # will fail. For this case, a static path is provided.
+        if handle is None and os.name == "posix" and sys.platform == "darwin":
+            handle = '/usr/lib/libc.dylib'
+
         if handle is not None:
-            cls.lib = ctypes.CDLL(handle)
+            try:
+                cls.lib = ctypes.CDLL(handle)
+            except OSError:
+                raise RuntimeError("Couldn't find `libc`'s `posix_memalign` to "
```
Comment (FabioLuporini) [inline devito/data/allocators.py:143]: no need to raise an error; just `cls.lib = None` and the error will be triggered, at a more suitable time (ie upon the first data allocation), by the next call to `_alloc_C_libcall`

**Pull request merged**